syntax = "proto3";

package DistributedRocksDB;

message SerializeByte {
	bytes buffer = 1;
}

message ReadResult{
    bytes value = 1;
    int32 err = 2;
}

message ReadRequest{
    uint32 key = 1;
    bool requirecache = 2;
    string clientidentifier = 3;
    string consistency = 4;
}

message WriteRequest{
    uint32 key = 1;
    bytes value = 2;
    string clientidentifier = 3;
    string consistency = 4;
}

message WriteResult{
    int32 err = 1;
}

message ClientCacheNotify {
    int32 key = 1;
}

message SubscribeForNotifications {
    string identifier = 1;
}

message Heartbeat {
    string msg = 1;
}

message RegisterRequest {
    string address = 1;
    int32 clusterid = 2;
}

message RegisterResult {
    string result = 1;
}

message SystemStateRequest {
    string request = 1;
}

message SystemState {
    string primary = 1;
    repeated string backups = 2;
    int32 keyrange = 3;
}

message SystemStateResult {
    repeated SystemState systemstate = 1;
}

message TxnFlushRequest {
    int32 numTxn = 1;
    int32 logIndex = 2;
}

message TxnFlushReply {
    int32 error = 1;
}

enum RecoveryRequestType {
    GET_TXNS = 0;
    REG_DONE = 1;
}

enum RecoveryReplyType {
    TXNS = 0;
    TXNS_DONE = 1;
}

message RecoverRequest {
    int32 lastLogIndex = 1;
    RecoveryRequestType reqType = 2;
}

message Txn{
    string key = 1;
    string value = 2;
}

message RecoverReply {
    int32 logIndex = 1;
    int32 error = 2;
    repeated Txn txns = 3;
    RecoveryReplyType replyType = 4;
}

service DistributedRocksDBService {
    rpc rpc_read(ReadRequest) returns (ReadResult) {}
    rpc rpc_write(WriteRequest) returns (WriteResult) {}
    rpc rpc_subscribeForNotifications(SubscribeForNotifications) returns (stream ClientCacheNotify) {}
    rpc rpc_unSubscribeForNotifications(SubscribeForNotifications) returns (SubscribeForNotifications) {}
    rpc rpc_heartbeat(stream SystemState) returns (Heartbeat) {}
    rpc rpc_clusterHeartbeat(stream SystemStateResult) returns (Heartbeat) {}
    rpc rpc_registerNewNode(RegisterRequest) returns (RegisterResult) {}
    rpc rpc_getSystemState(SystemStateRequest) returns (SystemState) {}
    rpc rpc_getClusterSystemState(SystemStateRequest) returns (SystemStateResult) {}
    rpc rpc_flush(TxnFlushRequest) returns (TxnFlushReply) {}
    rpc rpc_recover(stream RecoverRequest) returns (stream RecoverReply) {}
}